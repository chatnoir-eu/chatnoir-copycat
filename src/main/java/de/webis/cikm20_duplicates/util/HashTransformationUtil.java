package de.webis.cikm20_duplicates.util;

import lombok.experimental.UtilityClass;

@UtilityClass
public class HashTransformationUtil {

	public static int bytesToInt(byte[] bytes) {
		if(bytes == null || bytes.length != 4) {
			throw new IllegalArgumentException("Input must be not null and 4 bytes");
		}
		
		int ret = 0;
		
		for(int i=0; i<bytes.length; i++) {
			int current = 255 & (bytes[(bytes.length-1) -i]);
			ret = ret ^ (current<<(8*i));
		}
		
		return ret;
	}
	
	public static byte[] intToBytes(int actual, int i) {
		if(i != 4) {
			throw new IllegalArgumentException("I can only convert an int to 4 bytes");
		}
		
		return new byte[] {
			(byte)((actual>>24) & 255),
			(byte)((actual>>16) & 255),
			(byte)((actual>>8) & 255),
			(byte)((actual) & 255)
		};
	}
}
