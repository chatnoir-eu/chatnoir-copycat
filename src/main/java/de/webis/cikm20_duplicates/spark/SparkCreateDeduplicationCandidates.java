package de.webis.cikm20_duplicates.spark;

import java.io.Serializable;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.stream.Collectors;

import org.apache.curator.shaded.com.google.common.collect.Iterators;
import org.apache.htrace.shaded.fasterxml.jackson.databind.ObjectMapper;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.google.common.hash.BloomFilter;
import com.google.common.hash.Funnels;

import de.webis.cikm20_duplicates.util.FingerPrintUtil.Fingerprinter;
import de.webis.cikm20_duplicates.util.SourceDocuments.CollectionDocumentWithTopics;
import de.webis.cikm20_duplicates.util.SourceDocuments.DocumentWithFingerprint;
import de.webis.trec_ndd.trec_collections.AnseriniCollectionReader;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import io.anserini.collection.ClueWeb09Collection.Document;
import lombok.SneakyThrows;
import scala.Tuple2;

/**
 * 
 * @author Maik Fr√∂be
 *
 */
public class SparkCreateDeduplicationCandidates {

	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			JavaRDD<String> input = context.textFile("cikm2020/document-fingerprints");
			
//			JavaRDD<String> duplicationCandidates = duplicationCandidatesFromFingerprints(input)
//				.map(i -> i.toString());
//			
//			duplicationCandidatePairsFromFingerprints(duplicationCandidates)
//				.saveAsTextFile("cikm2020/candidate-pairs");
			
			hashPartitionToDocument(input, DeduplicationStrategy.MIN_HASH_DEDUPLICATION_STRATEGY)
				.aggregateByKey(0l, (count, doc) -> Long.valueOf((long)(count +1)), (i,j) -> Long.valueOf((long) i+j))
				.map(i -> "{\"hash-partition\": "+ i._1() +", \"count\": "+ i._2() +"}")
				.saveAsTextFile("cikm2020/count-per-hash-partition");
		}
	}
	
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/candidates");

		return new JavaSparkContext(conf);
	}
	
	public static JavaPairRDD<String, Iterable<Tuple2<String, CollectionDocument>>> topicsToImportantDocuments(JavaRDD<String> rdd) {
		JavaRDD<Tuple2<String, CollectionDocument>> ret = rdd
				.map(i -> CollectionDocumentWithTopics.fromString(i))
				.flatMap(i -> flattenTopicsForDoc(i));
 
		return ret.groupBy(i -> i._1());
	}

	private static Iterator<Tuple2<String, CollectionDocument>> flattenTopicsForDoc(CollectionDocumentWithTopics docWithTopics) {
		return docWithTopics.getTopics().stream()
				.map(topic -> new Tuple2<>(topic, docWithTopics.getDoc()))
				.iterator();
	}

	public static <T extends Comparable<T>> JavaRDD<Tuple2<String, Set<T>>> topicsToFingerPrintsOfImportantDocsPerTopic(JavaRDD<String> rdd, Fingerprinter<T> fingerprinter) {
		return topicsToImportantDocuments(rdd)
				.map(i -> new Tuple2<>(i._1, combineAll(i._2, fingerprinter)));
	}

	private static <T extends Comparable<T>> Set<T> combineAll(Iterable<Tuple2<String, CollectionDocument>> a, Fingerprinter<T> fingerprinter) {
		Set<T> ret = new HashSet<>();
		Iterator<CollectionDocument> docIterator = Iterators.transform(a.iterator(), i -> i._2());
		docIterator.forEachRemaining(doc -> ret.addAll(fingerprinter.fingerprint(doc)));

		return ret;
	}

	public static JavaRDD<Tuple2<String, CollectionDocument>> candidatesForAllSourceDocuments(JavaSparkContext context, JavaRDD<String> sourceDocuments, Fingerprinter<Integer> fingerprinter, AnseriniCollectionReader<Document> acr) {
		Map<String, BloomFilter<Integer>> topicBloomFilters = new HashMap<>();
		
		Map<String, Set<Integer>> topicToFingerPrintUnits = topicsToFingerPrintsOfImportantDocsPerTopic(sourceDocuments, fingerprinter)
				.collect().stream()
				.collect(Collectors.toMap(i -> i._1(), i -> i._2()));
		int overallElements = topicToFingerPrintUnits.values().stream().mapToInt(i -> i.size()).sum();
		
		BloomFilter<Integer> bf = BloomFilter.create(Funnels.integerFunnel(), overallElements + 100000, 1.0e-8);

		for(Entry<String, Set<Integer>> topicToFingerpintUnit : topicToFingerPrintUnits.entrySet()) {
			BloomFilter<Integer> topicBf = BloomFilter.create(Funnels.integerFunnel(), topicToFingerpintUnit.getValue().size() + 10000, 1.0e-8);
			
			for(Integer fingerprintUnit: topicToFingerpintUnit.getValue()) {
				topicBf.put(fingerprintUnit);
				bf.put(fingerprintUnit);
			}
			
			topicBloomFilters.put(topicToFingerpintUnit.getKey(), topicBf);
		}
		
		JavaRDD<CollectionDocument> allDocs = context.parallelize(acr.segmentPaths())
			.flatMap(s -> acr.collectionDocumentsInPath(s));
		
		return allDocs.flatMap(doc -> docToCandidateTopics(fingerprinter, doc, bf, topicBloomFilters));
	}
	
	private static <T extends Comparable<T>> Iterator<Tuple2<String, CollectionDocument>> docToCandidateTopics(Fingerprinter<T> fingerprinter, CollectionDocument doc, BloomFilter<T> bf, Map<String, BloomFilter<T>> topicBloomFilters) {
		List<T> fingerPrint = fingerprinter.fingerprint(doc);
		Set<String> candidateTopics = new HashSet<>();
		
		for(T fingerPrintUnit: fingerPrint) {
			if(bf.mightContain(fingerPrintUnit)) {
				for(Entry<String, BloomFilter<T>> e : topicBloomFilters.entrySet()) {
					if(e.getValue().mightContain(fingerPrintUnit)) {
						candidateTopics.add(e.getKey());
					}
				}
			}
		}
		
		return candidateTopics.stream()
				.map(i -> new Tuple2<>(i, doc))
				.iterator();
	}

	private static JavaPairRDD<Integer, String> hashPartitionToDocument(JavaRDD<String> docsWithFingerprint, DeduplicationStrategy f) {
		return docsWithFingerprint
				.map(i -> DocumentWithFingerprint.fromString(i))
				.flatMapToPair(doc -> extractHashesToDocId(doc, f));
	}
	
	public static JavaRDD<String> duplicationCandidatePairsFromFingerprints(JavaRDD<String> docsWithFingerprint, DeduplicationStrategy f) {
		JavaPairRDD<Integer, String> parsedInput = hashPartitionToDocument(docsWithFingerprint, f);
		
		return parsedInput.join(parsedInput)
				.map(i -> emitPairOrNull(i._2()))
				.filter(i -> i != null)
				.distinct();
	}
	
	public static List<Integer> useMinHash(DocumentWithFingerprint doc) {
		return doc.getMinHashParts();
	}
	
	public static List<Integer> useSimHash(DocumentWithFingerprint doc) {
		return doc.getSimHash65BitParts();
	}
	
	@SneakyThrows
	private static String emitPairOrNull(Tuple2<String, String> pair) {
		if(pair == null || pair._1() == null || pair._2() == null || pair._1().compareTo(pair._2()) >= 0) {
			return null;
		}

		Map<String, String> candidate = new LinkedHashMap<>();
		candidate.put("firstId", pair._1());
		candidate.put("secondId", pair._2());
				
		return new ObjectMapper().writeValueAsString(candidate);
	}

	private static Iterator<Tuple2<Integer, String>> extractHashesToDocId(DocumentWithFingerprint doc, DeduplicationStrategy f) {
		return f.extract(doc).stream()
				.map(hash -> new Tuple2<>(hash, doc.getDocId()))
				.iterator();
	}
	
	public static JavaRDD<DocumentWithFingerprint> duplicationCandidatesFromFingerprints(JavaRDD<String> docsWithFingerprint) {
		JavaRDD<DocumentWithFingerprint> parsedInput = docsWithFingerprint.map(i -> DocumentWithFingerprint.fromString(i));
		BloomFilter<Integer> bf = bf(parsedInput);
		
		return parsedInput.filter(doc -> doc.getMinHashParts().stream().anyMatch(i -> bf.mightContain(i)));
	}
	
	private static BloomFilter<Integer> bf(JavaRDD<DocumentWithFingerprint> docs) {
		List<Integer> allElements = docs.filter(i -> SparkCreateSourceDocuments.DOCS_TO_TOPIC.containsKey(i.getDocId()))
			.flatMap(i -> i.getMinHashParts().iterator())
			.distinct()
			.collect();
		
		BloomFilter<Integer> ret = BloomFilter.create(Funnels.integerFunnel(), allElements.size(), 1.0e-8);
		allElements.forEach(i -> ret.put(i));
		
		return ret;
	}
	
	static interface DeduplicationStrategy extends Serializable {
		public List<Integer> extract(DocumentWithFingerprint doc);
		
		@SuppressWarnings("serial")
		public static final DeduplicationStrategy MIN_HASH_DEDUPLICATION_STRATEGY = new DeduplicationStrategy() {
			@Override
			public List<Integer> extract(DocumentWithFingerprint doc) {
				return doc.getMinHashParts();
			}
		};
		
		@SuppressWarnings("serial")
		public static final DeduplicationStrategy SIM_HASH_DEDUPLICATION_STRATEGY = new DeduplicationStrategy() {
			@Override
			public List<Integer> extract(DocumentWithFingerprint doc) {
				return doc.getSimHash65BitParts();
			}
		};
	}
}
