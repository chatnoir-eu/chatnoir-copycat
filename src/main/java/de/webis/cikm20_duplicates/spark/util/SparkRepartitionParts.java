package de.webis.cikm20_duplicates.spark.util;

import java.util.LinkedList;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

public class SparkRepartitionParts {
	
	public static void main(String[] args) {
		//String inputDir = "cikm2020/canonical-link-graph/cc-2017-04-canonical-urls";
//		String inputDir = "cikm2020/canonical-link-graph/cc-2017-04-sample-0.1";
		String inputDir = "cikm2020/canonical-link-graph/cc-2015-11-sample-0.1";
		
		try (JavaSparkContext context = context()) {
			for(String suffix: createPartSuffixes()) {
				JavaRDD<String> input = context.textFile(inputDir +"/part*" + suffix);
				input.repartition(5).saveAsTextFile(inputDir + "-delete-me-tmp-"+ suffix);
			}
			
			JavaRDD<String> completeInput = context.textFile(inputDir + "-delete-me-tmp-*/part*");
			completeInput.repartition(10000).saveAsTextFile(inputDir + "-new");
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/repartition-data");
	
		return new JavaSparkContext(conf);
	}

	public static List<String> createPartSuffixes() {
		List<String> ret = new LinkedList<>();
		
		for(int i=0; i < 10; i++) {
			for(int j=0; j<10; j++) {
				ret.add(i + "" + j);
			}
		}
		
		return ret;
	}
}
