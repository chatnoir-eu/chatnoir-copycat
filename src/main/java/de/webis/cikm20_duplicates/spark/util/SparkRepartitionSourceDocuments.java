package de.webis.cikm20_duplicates.spark.util;

import org.apache.hadoop.io.compress.BZip2Codec;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

public class SparkRepartitionSourceDocuments {
	
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			String corpus = "cc-2015-11";
			JavaRDD<String> input = context.textFile("cikm2020/document-fingerprints-final/" + corpus +"*-jsonl");
			
			input.repartition(10000)
				.saveAsTextFile("cikm2020/document-fingerprints-final/" + corpus +"-jsonl.bzip2", BZip2Codec.class);
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/repartition-data");

		return new JavaSparkContext(conf);
	}
}
