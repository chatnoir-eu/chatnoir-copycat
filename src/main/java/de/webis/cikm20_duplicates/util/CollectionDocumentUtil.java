package de.webis.cikm20_duplicates.util;

import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.time.temporal.ChronoUnit;

import org.apache.commons.io.IOUtils;

import de.webis.WebisUUID;
import de.webis.cikm20_duplicates.spark.SparkEnrichRelevanceTransferPairs;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import io.anserini.index.transform.JsoupStringTransform;
import lombok.SneakyThrows;
import net.jodah.failsafe.Failsafe;
import net.jodah.failsafe.RetryPolicy;

public class CollectionDocumentUtil {

	@SneakyThrows
	public static void main(String[] args) {
//		// S3: 1.0 (Should be valid)
//		String firstId = "clueweb12-0001wb-23-26353", secondId = "clueweb12-0005wb-24-06245";
		
//		// S3: 0.0 (Should be invalid)
//		String firstId = "clueweb12-0912wb-53-11472", secondId = "clueweb12-1000tw-11-07961";
		
		// S3: 0.0 (Should be invalid)
//		String firstId = "clueweb12-0303wb-10-02360", secondId = "clueweb12-0715wb-02-10722";

//		// S3: 0.0 (Should be invalid)
//		String firstId = "clueweb12-0303wb-35-15203", secondId = "clueweb12-0816wb-90-18920";
		
		// S3: 1.0 (Should be valid)
//		String firstId = "clueweb12-0406wb-47-22712", secondId = "clueweb12-1305wb-37-22239";
		
//		// S3: 1.0 (Should be valid)
//		String firstId = "clueweb12-0308wb-07-19989", secondId = "clueweb12-1505wb-58-23691";
		
//		// S3: 0.899 (Should be valid)
//		String firstId = "clueweb12-0810wb-02-19724", secondId = "clueweb12-0811wb-20-10663";
		
		
//		// S3: 0.99 (Should be valid)
//		String firstId = "clueweb12-0807wb-51-02126", secondId = "clueweb12-0810wb-43-15854";
		
//		// S3: 0.79 (should be invalid)
//		String firstId = "clueweb12-1313wb-20-00368", secondId = "clueweb12-1313wb-24-32697";
		
//		// S3: 0.94 (should be valid)
//		String firstId = "clueweb12-0805wb-13-09655", secondId = "clueweb12-0809wb-63-03994";
		
//		// S3: 0.97 (should be valid)
//		String firstId = "clueweb12-0810wb-79-00932", secondId = "clueweb12-0811wb-54-18364";
		
//		// S3: 0.93 (should be valid)
//		String firstId = "clueweb12-1600tw-18-12106", secondId = "clueweb12-1900tw-30-12707";
		
//		// S3: 0.79 (should be invalid)
//		String firstId = "clueweb12-1018wb-50-14351", secondId = "clueweb12-1019wb-33-29258";
		
//		System.out.println("Retrieve " + firstId + " located at: " + chatNoirURL(firstId) +"&plain");
//		CollectionDocument a = loadCollectionDocument(firstId);

//		System.out.println("Retrieve " + secondId + " located at: " + chatNoirURL(secondId) +"&plain");
//		CollectionDocument b = loadCollectionDocument(secondId);
		
//		System.out.println("Calculate S3-Score between " + firstId + " and " + secondId);
//		System.out.println(SparkEnrichRelevanceTransferPairs.s3Score(a,b));
		
		// {"equivalentDocuments": ["<urn:uuid:59a4637d-976f-4a5c-9861-cd5f3fc6c737>","<urn:uuid:be66ee99-8ee6-46ce-9e96-de6067eff76a>","<urn:uuid:ecffb099-540a-4176-8a97-7152a0576383>"],"hash":[-244842496, 47039, 11927603, 14439680]}
		//String firstId = "<urn:uuid:59a4637d-976f-4a5c-9861-cd5f3fc6c737>";
		//String secondId = "<urn:uuid:be66ee99-8ee6-46ce-9e96-de6067eff76a>";
		
		//{"equivalentDocuments": ["<urn:uuid:84606826-df14-4f0d-94cc-7d828f880949>","<urn:uuid:a1a75ce9-255f-4acd-83f9-30afd4fa6631>"],"hash":[-889782272, 50871, 14287080, 7504640]}
		// String firstId = "<urn:uuid:84606826-df14-4f0d-94cc-7d828f880949>";
		// String secondId = "<urn:uuid:a1a75ce9-255f-4acd-83f9-30afd4fa6631>";
		
		//{"equivalentDocuments": ["<urn:uuid:aa833e3e-0033-4937-b499-14e26d972404>","<urn:uuid:fb1ec340-b003-4dad-aa02-9a3ebe34b75f>"],"hash":[1527775232, 344, 13172802, 9114880]}
		// String firstId = "<urn:uuid:aa833e3e-0033-4937-b499-14e26d972404>";
		// String secondId = "<urn:uuid:fb1ec340-b003-4dad-aa02-9a3ebe34b75f>";
		
		//{"equivalentDocuments": ["<urn:uuid:177caa19-fd12-48dd-b85f-aa9151e1048e>","<urn:uuid:c74bf9bb-c949-47a1-bd4c-967e528a3e0d>","<urn:uuid:fd3341db-5dbb-45fe-b5c3-7aa1f91527a6>"],"hash":[367919104, 13594, 1900705, 413184]}
		String firstId = "<urn:uuid:177caa19-fd12-48dd-b85f-aa9151e1048e>";
		String secondId = "<urn:uuid:fd3341db-5dbb-45fe-b5c3-7aa1f91527a6>";
		
		System.out.println(chatNoirURL("commoncrawl", firstId, "cc1704"));
		CollectionDocument a = loadCollectionDocument(firstId, new URL(chatNoirURL("commoncrawl", firstId, "cc1704")));
		System.out.println(chatNoirURL("commoncrawl", secondId, "cc1704"));
		CollectionDocument b = loadCollectionDocument(secondId, new URL(chatNoirURL("commoncrawl", secondId, "cc1704")));
		
		System.out.println(SparkEnrichRelevanceTransferPairs.s3Score(a,b));
	}
	
	public static final RetryPolicy<String> RETRY_FINAL = new RetryPolicy<String>()
			.handle(Exception.class)
			.withBackoff(3, 60, ChronoUnit.SECONDS)
			.withMaxRetries(3);	
	
	@SneakyThrows
	public static CollectionDocument loadCollectionDocument(String documentId) {
		return loadCollectionDocument(new URL(chatNoirURL(documentId)), documentId, RETRY_FINAL);
	}
	

	public static CollectionDocument loadCollectionDocument(String documentId, URL chatNoirUrl) {
		return loadCollectionDocument(chatNoirUrl, documentId, RETRY_FINAL);
	}
	
	public static CollectionDocument loadCollectionDocument(URL url, String documentId, RetryPolicy<String> retryPolicy) {
		String documentContent = loadRawDocumentFromChatnoir(url, retryPolicy);
		
		JsoupStringTransform stringTransform = new JsoupStringTransform();
		return CollectionDocument.collectionDocument(stringTransform.apply(documentContent), documentId);
	}
	
	public static String loadRawDocumentFromChatnoir(URL url, RetryPolicy<String> retryPolicy) {
		return Failsafe.with(retryPolicy).get(() -> IOUtils.toString(url, StandardCharsets.UTF_8));
	}

	private static String webisUUID(String documentId) {
		return webisUUID(longChatNoirId(documentId), documentId);
	}
	
	private static String webisUUID(String prefix, String documentId) {
		return new WebisUUID(prefix)
				.generateUUID(documentId).toString();
	}
	
	private static String longChatNoirId(String documentId) {
		if(documentId.startsWith("clueweb09")) {
			return "clueweb09";
		} else if (documentId.startsWith("clueweb12")) {
			return "clueweb12";
		}
		
		throw new RuntimeException("ID '" + documentId + "' is not supported.");
	}
	
	private static String shortChatNoirId(String documentId) {
		if (documentId.startsWith("clueweb09")) {
			return "cw09";
		} else if (documentId.startsWith("clueweb12")) {
			return "cw12";
		}
		
		throw new RuntimeException("ID '" + documentId + "' is not supported.");
	}
	
	public static String chatNoirURL(String documentId) {
		return "https://chatnoir.eu/cache?uuid=" + webisUUID(documentId) + "&index="+ shortChatNoirId(documentId) +"&raw";
	}
	
	public static String chatNoirURL(String prefix, String documentId, String index) {
		return "https://chatnoir.eu/cache?uuid=" + webisUUID(prefix, documentId) + "&index="+ index +"&raw";
	}
}
