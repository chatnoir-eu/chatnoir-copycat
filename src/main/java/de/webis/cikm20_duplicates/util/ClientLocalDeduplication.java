package de.webis.cikm20_duplicates.util;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.curator.shaded.com.google.common.collect.Iterators;
import org.apache.htrace.shaded.fasterxml.jackson.databind.ObjectMapper;

import com.google.common.collect.ImmutableList;

import de.aitools.ir.fingerprinting.representer.Hash;
import de.webis.cikm20_duplicates.spark.SparkCreateDeduplicationCandidates.DeduplicationUnit;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;
import scala.Tuple2;

public class ClientLocalDeduplication {

	public static List<String> dedup(Iterable<Tuple2<Integer, DeduplicationUnit>> group) {
		List<String> ret = new ArrayList<>();
		Map<String, List<Tuple2<Integer, DeduplicationUnit>>> equals = sortedList(group);

		for(List<Tuple2<Integer, DeduplicationUnit>> equal: equals.values()) {
			ret.addAll(equalPairs(equal));
		}
		
		for(List<DeduplicationUnit> toDedup: deduplicationPairs(equals)) {
			for(String bla : fullDeduplication(toDedup)) {
				if(!ret.contains(bla)) {
					ret.add(bla);
				}
			}
		}
		
		return ret;
	}
	
	public static Collection<List<DeduplicationUnit>> deduplicationPairs(Map<String, List<Tuple2<Integer, DeduplicationUnit>>> equals) {
		Map<Integer, List<DeduplicationUnit>> ret = new LinkedHashMap<>();
		
		if(equals.size() < 100000) {
			List<DeduplicationUnit> tmpRet = new ArrayList<>();
			for(List<Tuple2<Integer, DeduplicationUnit>> equal: equals.values()) {
				if(equal.size() < 1) {
					continue;
				}
				
				tmpRet.add(equal.get(0)._2);
			}
			
			return Arrays.asList(tmpRet);
		}
		
		for(List<Tuple2<Integer, DeduplicationUnit>> equal: equals.values()) {
			if(equal.size() < 1) {
				continue;
			}
			
			Tuple2<Integer, DeduplicationUnit> i = equal.get(0);
			List<Integer> rekursiveInts = HashTransformationUtil.removeIntFromUnderlyingBitArray(i._2.getHashParts(), i._1());
			for(int recursivePart: rekursiveInts) {
				if(!ret.containsKey(recursivePart)) {
					ret.put(recursivePart, new ArrayList<>());
				}
				
				ret.get(recursivePart).add(i._2);
			}
		}
		
		return ret.values();
	}
	
	public static List<String> fullDeduplication(List<DeduplicationUnit> orig) {
		List<String> ret = new LinkedList<>();
		List<Tuple2<String, byte[]>> idToHash = new ArrayList<>(orig.size());
		
		for(DeduplicationUnit d: orig) {
			idToHash.add(new Tuple2<>(d.getId(), HashTransformationUtil.integersToHash(d.getHashParts())));
		}
		
		for(int i=0; i<idToHash.size(); i++) {
			Tuple2<String, byte[]> left = idToHash.get(i);

			for(int j=i+1; j< idToHash.size(); j++) {
				Tuple2<String, byte[]> right = idToHash.get(j);
				
				int hemming = Hash.getHammingDistance(left._2(), right._2());
				if(hemming <= 3) {
					ret.add(nearDuplicate(left._1(), right._1(), hemming));
				}
			}
		}
		
		return ret;
	}

	private static Collection<String> equalPairs(List<Tuple2<Integer, DeduplicationUnit>> equal) {
		List<String> ret = new ArrayList<>();
		
		for(int i= 1; i< equal.size(); i++) {
			ret.add(nearDuplicate(equal.get(0)._2.getId(), equal.get(i)._2.getId(), 0));
		}
		
		return ret;
	}
	
	private static String nearDuplicate(String id1, String id2, int hemming) {
		return "{\"firstId\":\"" + id1 + "\",\"secondId\":\"" + id2 + "\",\"hemmingDistance\":"+ hemming +"}";
	}


	private static void failIfElementsAreNotAllFromTheSameGroup(Iterable<Tuple2<Integer, DeduplicationUnit>> group) {
		Set<Integer> ret = new HashSet<>();
		Iterators.transform(group.iterator(), i -> i._1()).forEachRemaining(i -> ret.add(i));
		
		if(ret.size() > 1) {
			throw new IllegalArgumentException("");
		}
	}
	
	public static Map<String, List<Tuple2<Integer, DeduplicationUnit>>> sortedList(Iterable<Tuple2<Integer, DeduplicationUnit>> group) {
		List<Tuple2<Integer, DeduplicationUnit>> tmp = ImmutableList.copyOf(group.iterator());
		tmp = new ArrayList<>(tmp);
		Collections.sort(tmp, (a,b) -> a._2.getId().compareTo(b._2.getId()));
		failIfElementsAreNotAllFromTheSameGroup(tmp);
		
		Map<String, List<Tuple2<Integer, DeduplicationUnit>>> ret = new LinkedHashMap<>();
		for(Tuple2<Integer, DeduplicationUnit> t : tmp) {
			String key = t._2().getHashParts().toString();
			
			if(!ret.containsKey(key)) {
				ret.put(key, new LinkedList<>());
			}
			
			ret.get(key).add(t);
		}
		
		
		return ret;
	}
	
	@Data
	@NoArgsConstructor
	@AllArgsConstructor
	public static class DeduplicationTask {
		private List<DeduplicationUnit> entries;
		
		@Override
		@SneakyThrows
		public String toString() {
			return new ObjectMapper().writeValueAsString(this);
		}
		
		@SneakyThrows
		public static DeduplicationTask fromString(String value) {
			return new ObjectMapper().readValue(value, DeduplicationTask.class);
		}
	}

	public static java.util.Iterator<String> workingPackages(Iterable<Tuple2<Integer, DeduplicationUnit>> group) {
		Map<String, List<Tuple2<Integer, DeduplicationUnit>>> equals = sortedList(group);
		Collection<List<DeduplicationUnit>> ret = deduplicationPairs(equals);
		
		return com.google.common.collect.Iterators.transform(
			ret.iterator(), 
			i -> new DeduplicationTask(i).toString()
		);
	}
}
