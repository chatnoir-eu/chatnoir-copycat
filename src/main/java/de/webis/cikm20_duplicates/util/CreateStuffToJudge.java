package de.webis.cikm20_duplicates.util;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.databind.ObjectMapper;

import de.webis.cikm20_duplicates.spark.SparkCalculateCanonicalLinkGraphEdgeLabels.CanonicalLinkGraphEdge2;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import lombok.SneakyThrows;

public class CreateStuffToJudge {
	private static final String
		DIR = "src/main/resources/canonical-link-edges-human-judgments/",
		
		//CREATED BY: de.webis.cikm20_duplicates.spark.eval.SparkSampleS3EdgesPerBin
		INPUT = DIR + "cw12-sampled-edges.jsonl";
	
	@SneakyThrows
	public static void main(String[] args) {
		for(String line: Files.readAllLines(Paths.get(INPUT))) {
			persistExamples(line);
		}
	}
	
	@SneakyThrows
	@SuppressWarnings({ "unchecked", "rawtypes" })
	private static void persistExamples(String json) {
		Map<String, Object> m = new ObjectMapper().readValue(json, Map.class); 
		
		if(!m.containsKey("s3Bucket") || m.get("s3Bucket") == null) {
			return;
		} else {
			System.out.println("Bucket: " + m.get("s3Bucket"));
		}
		
		List edges = (List) m.get("sample-edges");
		edges = edges.subList(0, 10);
		
		for(Object e: edges) {
			String eStr = new ObjectMapper().writeValueAsString(e);
			CanonicalLinkGraphEdge2 eVal = CanonicalLinkGraphEdge2.fromString(eStr);
			
			persistExample(eVal);
		}
	}
	
	@SneakyThrows
	private static void persistExample(CanonicalLinkGraphEdge2 bla) {
		CollectionDocument src = bla.getFirstDoc().getDoc();
		CollectionDocument target = bla.getSecondDoc().getDoc();
		
		Path dir = Paths.get(DIR + "judgments-by-daniel/" + src.getId() + "__to__" + target.getId());
		dir.toFile().mkdirs();
		
		Files.write(dir.resolve("metadata"), metadata(src, target, bla.getCanonicalLink(), bla.getS3score()).getBytes(StandardCharsets.UTF_8));
		Files.write(dir.resolve("src-doc"), src.getFullyCanonicalizedContent().getBytes(StandardCharsets.UTF_8));
		Files.write(dir.resolve("target-doc"), target.getFullyCanonicalizedContent().getBytes(StandardCharsets.UTF_8));
	}
	
	private static String metadata(CollectionDocument a, CollectionDocument b, String canonicalUrl, double s3) {
		return "Canonical-Link: " + canonicalUrl +"\n" +
				"First: "+ CollectionDocumentUtil.chatNoirURL(a.getId()) +"&plain\n" + 
				"Second: "+ CollectionDocumentUtil.chatNoirURL(b.getId()) +"&plain\n" +
				"S3: " + s3 +"\n";
	}
}
