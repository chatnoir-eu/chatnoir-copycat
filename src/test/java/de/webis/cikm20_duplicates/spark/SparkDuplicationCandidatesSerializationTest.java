package de.webis.cikm20_duplicates.spark;

import java.util.Arrays;
import java.util.List;

import org.approvaltests.Approvals;
import org.junit.Test;

import com.holdenkarau.spark.testing.SharedJavaSparkContext;

import de.webis.cikm20_duplicates.spark.SparkCreateDeduplicationCandidates.DeduplicationUnit;
import de.webis.cikm20_duplicates.util.ClientLocalDeduplication;
import de.webis.cikm20_duplicates.util.ClientLocalDeduplicationTest;
import scala.Tuple2;

public class SparkDuplicationCandidatesSerializationTest extends SharedJavaSparkContext {

	@Test
	public void testWithEmptyDocuments() {
		Iterable<Tuple2<Integer, DeduplicationUnit>> input = ClientLocalDeduplicationTest.firstLargeDeduplicationInput();
		
		List<String> actual = SparkCreateSourceDocumentsIntegrationTest.sorted(jsc().parallelize(Arrays.asList(input))
			.flatMap(i -> ClientLocalDeduplication.workingPackages(i)));
	
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	public void testWithSecondExample() {
		Iterable<Tuple2<Integer, DeduplicationUnit>> input = ClientLocalDeduplicationTest.secondLargeDeduplicationInput();
		
		List<String> actual = SparkCreateSourceDocumentsIntegrationTest.sorted(jsc().parallelize(Arrays.asList(input))
			.flatMap(i -> ClientLocalDeduplication.workingPackages(i)));
	
		Approvals.verifyAsJson(actual);
	}
}
