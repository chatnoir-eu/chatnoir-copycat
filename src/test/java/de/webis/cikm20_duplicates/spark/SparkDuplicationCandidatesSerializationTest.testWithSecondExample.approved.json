[
  "{\"entries\":[{\"id\":\"1\",\"hashParts\":[16842752,257,65537,66048]},{\"id\":\"10\",\"hashParts\":[16842752,257,65537,66304]},{\"id\":\"12\",\"hashParts\":[16842752,257,65537,328448]},{\"id\":\"5\",\"hashParts\":[16842752,257,65537,65792]},{\"id\":\"7\",\"hashParts\":[257,100,23,16]},{\"id\":\"8\",\"hashParts\":[257,21,41,18]},{\"id\":\"9\",\"hashParts\":[257,31,97,255]},{\"id\":\"999\",\"hashParts\":[16842752,257,65542,328448]}]}"
]